### PR TITLE
Add `NoCredentialsCache` that offers no caching ability

### PR DESCRIPTION
**Pull request by ysaito1001**
## Motivation and Context
Related to https://github.com/awslabs/aws-sdk-rust/issues/809

## Description
It has been discovered that when `AssumeRoleProvider` is used, the Rust SDK emits `credentials cache miss occurred` twice per request. The reason why that log is shown twice is illustrated in the following diagram:

Screenshot 2023-05-19 at 4 10 20 PM

One of the cache miss messages is due to the fact `AssumeRoleProvider` internally uses an STS client, which, in turn, is wrapped by a `LazyCredentialsCache` by default. However, that use of `LazyCredentialsCache` is pointless because caching is already in effect with the outermost `LazyCredentialsCache`.

This PR adds a new kind of `CredentialsCache`, `NoCredentialsCache`. As its name suggests, it simplify delegates `provide_cached_credentials` to the underlying provider's `provide_credentials` with no caching functionality. We then update `SsoCredentialsProvider`, `AssumeRoleProvider`, and `WebIdentityTokenCredentialsProvider` to use `NoCredentialsCache` for their STS clients so the logs won't show `credentials cache miss occurred` twice per request.

## Testing
- Added unit tests for `NoCredentialsCache`
- Updated unit test for `AssumeRoleProvider` to verify `NoCredentialsCache` is used by default

## Checklist
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

**Comment by rcoh** [comment]
did not closely review PR–I think I prefer something like `Uncached` or `NotCached`, `NoCredentialsCache` sounds like the absence of something?

**Comment by ysaito1001** [comment]
> did not closely review PR–I think I prefer something like `Uncached` or `NotCached`, `NoCredentialsCache` sounds like the absence of something?

Yeah, naming is hard. I unnecessarily tried sticking to `LazyCredentialsCache` and only replaced `Lazy` with `No`. Although `NoCredentialsCache` is a `pub(crate)`, I agree that it's worth giving it a proper name.

**Review by jdisanti — APPROVED** [review]
Looks good!

File: aws/rust-runtime/aws-config/src/sts/assume_role.rs [hunk 1]
```diff
@@ -180,6 +180,9 @@ impl AssumeRoleProviderBuilder {
     }
 
     /// Set the [`CredentialsCache`] for credentials retrieved from STS.
+    ///
+    /// By default, an [`AssumeRoleProvider`] internally uses 'NoCredentialsCache` because the
+    /// provider itself will be wrapped by `LazyCredentialsCache` when a service client is created.
     pub fn credentials_cache(mut self, cache: CredentialsCache) -> Self {
```
[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws/rust-runtime/aws-config/src/sts/assume_role.rs:186]: It's reasonable to disallow this configuration. The only reason I hesitated to remove it is that if I did, that'd be a breaking change. Seeing your comment, however, I feel more comfortable removing it, especially when the other providers' builders do not have the `credentials_cache` method.

[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-config/src/sts/assume_role.rs:186]: We could just mark it as `#[deprecated]` for now and remove it in the future.

**Pull request merged**